core/chains/evm/client/node_selector_priority_level.go

```diff
@@ -0,0 +1,103 @@
+package client
+
+import (
+	"math"
+	"sort"
+	"sync/atomic"
+)
+
+type priorityLevelNodeSelector struct {
+	nodes           []Node
+	roundRobinCount []atomic.Uint32
+}
+
+type nodesWithPriority struct {
+	node     Node
+	priority int32
+}
+
+func NewPriorityLevelNodeSelector(nodes []Node) NodeSelector {
+	return &priorityLevelNodeSelector{
+		nodes:           nodes,
+		roundRobinCount: make([]atomic.Uint32, nrOfPriorityTiers(nodes)),
+	}
+}
+
+func (s priorityLevelNodeSelector) Select() Node {
+	nodes := s.getHighestPriorityAliveTier()
+
+	if len(nodes) == 0 {
+		return nil
+	}
+	priorityLevel := nodes[len(nodes)-1].priority
+
+	// NOTE: Inc returns the number after addition, so we must -1 to get the "current" counter
+	count := s.roundRobinCount[priorityLevel].Add(1) - 1
+	idx := int(count % uint32(len(nodes)))
+
+	return nodes[idx].node
+}
+
+func (s priorityLevelNodeSelector) Name() string {
+	return NodeSelectionMode_PriorityLevel
+}
+
+// getHighestPriorityAliveTier filters nodes that are not in state NodeStateAlive and
+// returns only the highest tier of alive nodes
+func (s priorityLevelNodeSelector) getHighestPriorityAliveTier() []nodesWithPriority {
+	var nodes []nodesWithPriority
+	for _, n := range s.nodes {
+		if n.State() == NodeStateAlive {
+			nodes = append(nodes, nodesWithPriority{n, n.Order()})
+		}
+	}
+
+	if len(nodes) == 0 {
+		return nil
+	}
+
+	return removeLowerTiers(nodes)
+}
+
+// removeLowerTiers take a slice of nodesWithPriority and keeps only the highest tier
+func removeLowerTiers(nodes []nodesWithPriority) []nodesWithPriority {
+	sort.Slice(nodes, func(i, j int) bool {
```

Comment:
do we want a stable sort here? https://pkg.go.dev/sort#SliceStable

Reply:
Yeah, stable sort makes sense in this scenario, changed in [`2c3b889` (#9329)](https://github.com/smartcontractkit/chainlink/pull/9329/commits/2c3b889363a9a0274b0163340157234ab4b86162)